ops/framework.py

```diff
@@ -489,6 +489,13 @@ def __repr__(self):
         event_kinds = ', '.join(sorted(self._event_kinds()))
         return f'<{k.__module__}.{k.__qualname__}: {event_kinds}>'
 
+    def __getattr__(self, name: str):
+        """This tells type checkers to suspend judgement on this type's attributes.
+
+        This allows charms to freely write `self.on.foobar_baz` without mypy/pyright whining.
+        """
+        return super().__getattribute__(name)
```

Comment:
I checked this out locally and replaced this line with `raise AttributeError`, and all the tests still passed, so this code path isn't being exercised. That surprised me, but I guess it's because when we access such dynamically-defined events they *are* defined on the class and so `__getattr__` never gets called. Does that match your thinking?